src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        /// <summary>
+        /// Split the elements of a sequence into chunks of size at most <paramref name="maxSize"/>.
+        /// </summary>
+        /// <remarks>
+        /// Every chunk except the last will be of size <paramref name="maxSize"/>.
+        /// The last chunk will contain the remaining elements and may be of a smaller size.
+        /// </remarks>
+        /// <param name="source">
+        /// An <see cref="IEnumerable{T}"/> whose elements to chunk.
+        /// </param>
+        /// <param name="maxSize">
+        /// Maximum size of each chunk.
+        /// </param>
+        /// <typeparam name="TSource">
+        /// The type of the elements of source.
+        /// </typeparam>
+        /// <returns>
+        /// An <see cref="IEnumerable{T}"/> that contains the elements the input sequence split into chunks of size <paramref name="maxSize"/>.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="source"/> is null.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="maxSize"/> is below 1.
+        /// </exception>
+        public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> source, int maxSize)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (maxSize < 1)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.maxSize);
+            }
+
+            return source switch
+            {
+                IReadOnlyCollection<TSource> collection => ChunkIteratorOptimized(source, maxSize, collection.Count),
+                ICollection<TSource> collection => ChunkIteratorOptimized(source, maxSize, collection.Count),
```

Comment:
I don't think it's worth having an entirely separate implementation of the iterator just to properly size the last array yielded; the whole design of this method is that it allocates a new array every chunk, so it's in no way a high-performance method.  If you really want to do that, you could just have the one iterator accept an additional optional argument.

Reply:
(Also, even if we kept something along these lines, it's not valid to check the Count in the Chunk method itself.  The returned enumerable needs to be usable / reiterable many times, and the count could change between GetEnumerator calls.  If, for example, the collection was empty when Chunk was called, then an item was added, and then the enumerable returned from Chunk was enumerated, it needs to include the added item.)

Reply:
> I don't think it's worth having an entirely separate implementation of the iterator just to properly size the last array yielded; the whole design of this method is that it allocates a new array every chunk, so it's in no way a high-performance method. If you really want to do that, the one iterator method itself could type test, check the count, and factor it in if relevant. But at that point, I think we'd want to take it further, specializing instead for `IList<T>`, and also taking advantage of its indexer.)

Thanks for the idea of using the  `IList<T>` indexer. After doing some benchmarking, I noticed that knowing the `count` of `ICollection<T>`  did almost nothing.

I did try the indexer way instead and came up with the following result from my benchmarks:
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.330 (2004/?/20H1)
Intel Xeon CPU E3-1505M v5 2.80GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.102
  [Host]        : .NET Core 5.0.2 (CoreCLR 5.0.220.61120, CoreFX 5.0.220.61120), X64 RyuJIT
  .NET Core 5.0 : .NET Core 5.0.2 (CoreCLR 5.0.220.61120, CoreFX 5.0.220.61120), X64 RyuJIT

Job=.NET Core 5.0  Runtime=.NET Core 5.0  

```
|         Method | Length | ChunkSize |        Mean |     Error |    StdDev |
|--------------- |------- |---------- |------------:|----------:|----------:|
|           **List** |     **10** |         **3** |   **194.47 ns** |  **3.813 ns** |  **4.683 ns** |
|  ListOptimized |     10 |         3 |   122.72 ns |  1.159 ns |  1.084 ns |
|          Array |     10 |         3 |   161.04 ns |  1.603 ns |  1.421 ns |
| ArrayOptimized |     10 |         3 |   130.32 ns |  1.280 ns |  1.197 ns |
|           **List** |     **10** |         **5** |   **150.48 ns** |  **2.099 ns** |  **1.861 ns** |
|  ListOptimized |     10 |         5 |    91.85 ns |  0.685 ns |  0.607 ns |
|          Array |     10 |         5 |   118.84 ns |  1.412 ns |  1.321 ns |
| ArrayOptimized |     10 |         5 |    99.11 ns |  1.481 ns |  1.156 ns |
|           **List** |     **10** |        **15** |   **154.36 ns** |  **1.205 ns** |  **1.006 ns** |
|  ListOptimized |     10 |        15 |    76.54 ns |  0.759 ns |  0.672 ns |
|          Array |     10 |        15 |   115.75 ns |  1.457 ns |  1.363 ns |
| ArrayOptimized |     10 |        15 |    78.71 ns |  0.781 ns |  0.693 ns |
|           **List** |    **100** |         **3** | **1,352.16 ns** | **16.573 ns** | **14.692 ns** |
|  ListOptimized |    100 |         3 |   826.53 ns | 12.797 ns | 11.971 ns |
|          Array |    100 |         3 | 1,064.71 ns | 16.106 ns | 12.574 ns |
| ArrayOptimized |    100 |         3 |   866.36 ns |  5.925 ns |  5.542 ns |
|           **List** |    **100** |         **5** | **1,121.41 ns** | **17.625 ns** | **16.487 ns** |
|  ListOptimized |    100 |         5 |   611.66 ns |  3.622 ns |  3.211 ns |
|          Array |    100 |         5 |   855.31 ns |  8.208 ns |  7.276 ns |
| ArrayOptimized |    100 |         5 |   654.26 ns |  8.101 ns |  7.181 ns |
|           **List** |    **100** |        **15** | **1,011.28 ns** | **11.540 ns** | **10.794 ns** |
|  ListOptimized |    100 |        15 |   423.88 ns |  4.180 ns |  3.705 ns |
|          Array |    100 |        15 |   726.99 ns |  5.472 ns |  4.851 ns |
| ArrayOptimized |    100 |        15 |   458.76 ns |  6.211 ns |  5.506 ns |

I could include the new solution if the performance is worth the added complexity.

Reply:
> (Also, even if we kept something along these lines, it's not valid to check the Count in the Chunk method itself. The returned enumerable needs to be usable / reiterable many times, and the count could change between GetEnumerator calls. If, for example, the collection was empty when Chunk was called, then an item was added, and then the enumerable returned from Chunk was enumerated, it needs to include the added item.)

Thanks for noticing that flaw, I have fixed that issue and also added tests to cover this scenario.